Haystack-OPEA integration

danielfleischer opened the pull request:
Work done to contribute the integration, now fits in a dedicate repository.

danielfleischer committed:
Haystack-OPEA integration

Work done to contribute the integration, now fits in a dedicate repository.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Missing init file

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

rbrugaro commented on haystack_opea/utils/opea_backend.py:
```diff
@@ -0,0 +1,99 @@
+    def generate(self, prompt: str) -> Tuple[List[str], List[Dict[str, Any]]]:
+        url = f"{self.api_url}/v1/chat/completions"
+
+        try:
```
This fails with models served with vllm. Because vllm supports multiple models simultaneously it requires model ("model": "meta-llama/Llama-3.3-70B-Instruct",) to be provided. TGI I believe if you are using just one model it can be omitted. Please lets make it so it works for both serving frameworks

danielfleischer replied:
Added a `model_id` attribute throughout, default to empty string, see d272150e36dc349c99e98b0cb75f995ec37d0097.

rbrugaro commented on haystack_opea/utils/opea_backend.py:
```diff
@@ -0,0 +1,99 @@
+        return embedding, meta
+
+    def generate(self, prompt: str) -> Tuple[List[str], List[Dict[str, Any]]]:
+        url = f"{self.api_url}/v1/chat/completions"
```
I provided url as "http://localhost:8086"  but url became "http://localhost:8086/v1/v1/chat/completions". to make it run i removed the /v1 but we need to check if api_url somewhere is already appending v1.

danielfleischer replied:
Should be fixed by 317b440.

rbrugaro commented:
thanks @danielfleischer for the integration! 
for the LangChain-OPEA there is a [CI test](https://github.com/opea-project/LangChain-OPEA/blob/main/tests/test_integration_langchain_opea.sh) that launches the OPEA microservice for embedding and llm serving and validates the package. I can see there is a PR open to enable CI so not sure if its live but I think we should do the same here.

danielfleischer committed:
Extra v1 in path removed

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Re-introducing CI integration test script

TODO: need to add GitHub action to run it.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer commented:
I added the integration test script in 49784e9. Need to add github action calling it.

danielfleischer committed:
feat(generator): add model_id parameter to OPEAGenerator and OPEABackend classes

+ Added ~model_id~ parameter to ~OPEAGenerator~ and ~OPEABackend~ classes.
+ Updated tests to include ~model_id~ parameter.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

rbrugaro commented:
@chensuyue can you help set up the CI? @danielfleischer already added the test script

ftian1 commented:
@danielfleischer  one of CI tests fails, pls check it

chensuyue commented:
> @chensuyue can you help set up the CI? @danielfleischer already added the test script

Sure, I will follow this.

lvliang-intel commented:
@danielfleischer @chensuyue,
please check the CI issue.

cd /home/runner/work/Haystack-OPEA/Haystack-OPEA
  fail="FALSE"
  repo_name=opea-project/Haystack-OPEA
  if [ "$(echo "$repo_name"|cut -d'/' -f1)" != "opea-project" ]; then
    owner=$(echo "opea-project/Haystack-OPEA" |cut -d'/' -f1)
    branch="https://github.com/$owner/Haystack-OPEA/tree/changes"
  else
    branch="https://github.com/opea-project/Haystack-OPEA/blob/changes"
  fi
  link_head="https://github.com/opea-project/Haystack-OPEA/blob/main"
  
  merged_commit=$(git log -1 --format='%H')
  changed_files="$(git diff --name-status --diff-filter=ARM [3](https://github.com/opea-project/Haystack-OPEA/actions/runs/14107058574/job/39515930671?pr=1#step:4:3)39a459b91687303083[4](https://github.com/opea-project/Haystack-OPEA/actions/runs/14107058574/job/39515930671?pr=1#step:4:4)a7fec42e229a35ba3e21 ${merged_commit} | awk '/\.md$/ {print $NF}')"
  png_lines=$(grep -Eo '\]\([^)]+\)' --include='*.md' -r .|grep -Ev 'http')
  if [ -n "$png_lines" ]; then
    for png_line in $png_lines; do
      refer_path=$(echo "$png_line"|cut -d':' -f1 | cut -d'/' -f2-)
      png_path=$(echo "$png_line"|cut -d '(' -f2 | cut -d ')' -f1)
  
      if [[ "${png_path:0:1}" == "/" ]]; then
        check_path=$png_path
      elif [[ "$png_path" == *#* ]]; then
        relative_path=$(echo "$png_path" | cut -d '#' -f1)
        if [ -n "$relative_path" ]; then
          check_path=$(dirname "$refer_path")/$relative_path
          png_path=$(echo "$png_path" | awk -F'#' '{print "#" $2}')
        else
          check_path=$refer_path
        fi
      else
        check_path=$(dirname "$refer_path")/$png_path
      fi
  
      if [ -e "$check_path" ]; then
        real_path=$(realpath $check_path)
        if [[ "$png_line" == *#* ]]; then
          if [ -n "changed_files" ] && echo "$changed_files" | grep -q "^${refer_path}$"; then
            url_dev=$branch$(echo "$real_path" | sed 's|.*/Haystack-OPEA||')$png_path
            response=$(curl -I -L -s -o /dev/null -w "%{http_code}" "$url_dev")
            if [ "$response" -ne [20](https://github.com/opea-project/Haystack-OPEA/actions/runs/14107058574/job/39515930671?pr=1#step:4:20)0 ]; then
              echo "**********Validation failed, try again**********"
              response_retry=$(curl -s -o /dev/null -w "%{http_code}" "$url_dev")
              if [ "$response_retry" -eq 200 ]; then
                echo "*****Retry successfully*****"
              else
                echo "Invalid path from /home/runner/work/Haystack-OPEA/Haystack-OPEA/$refer_path: $png_path, link: $url_dev"
                fail="TRUE"
              fi
            else
              echo "Validation succeed $png_line"
            fi
          fi
        fi
      else
        echo "$check_path does not exist"
        fail="TRUE"
      fi
    done
  fi
  
  if [[ "$fail" == "TRUE" ]]; then
    exit 1
  else
    echo "All hyperlinks are valid."
  fi
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
Error: Process completed with exit code 1.

danielfleischer committed:
Trying to debug GH action

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer commented:
Fixed. There was a `grep` command that didn't find anything, which is fine but it exit with code 1 which kills a github action.

danielfleischer commented:
Thanks everyone!
